arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+{
+  volatile uint32_t error_passive;     /* Changes to error passive count */
+  volatile uint32_t bus_off;           /* Changes to bus off count */
+  volatile uint32_t arbitration_loss;  /* Arbitration loss errors count */
+  volatile uint32_t rx_overload;       /* Rx overload errors count */
+  volatile uint32_t bit_errors;        /* Bit errors count */
+  volatile uint32_t stuff_errors;      /* Stuffing errors count */
+  volatile uint32_t ack_errors;        /* Ack errors count */
+  volatile uint32_t form_errors;       /* Form errors count */
+  volatile uint32_t crc_errors;        /* CRC errors count */
+  volatile uint32_t stuck_at_0;        /* Stuck at 0 errors count */
+  volatile uint32_t restarts;          /* CAN controller re-starts count */
+  volatile uint32_t txb_sent;          /* Tx messages sent count */
+} mpfs_can_device_stats_t;
+
+/* The structure mpfs_can_instance_t is used by the driver to manage the
+ * configuration and operation of each MSS CAN peripheral. The instance
+ * content should only be accessed by using the respective API functions.
+ *
+ * Each API function has a pointer to this instance as first argument.
+ */
+
+typedef struct can_instance
+{
+  uintptr_t reg_base;       /* Pointer to CAN base register address */
+
+  uint32_t bitrate_value;    /* The numerical bitrate value in bit/s */
+
+  bool bifup;               /* Indicates if the CAN is up or down. */
+
+  /* Interrupt handling */
+
+  uint8_t irqn;             /* IRQ number */
+  uint32_t isr;             /* Interrupt status register */
+
+  /* Error status and Stats */
+
+  uint8_t error_status;                 /* Error status */
+  uint32_t tx_err_count;                /* Tx error count */
+  uint32_t rx_err_count;                /* Rx error count */
+  mpfs_can_device_stats_t stats;        /* Device statistics */
+
+  /* buffer count */
+
+  uint8_t  basic_can_rxb_count; /* number of rx buffers */
+  uint8_t  basic_can_txb_count; /* number of tx buffers */
+
+  /* Frame descriptors */
+
+  struct can_frame *txdesc; /* Pointer to the transmit frame descriptor. */
+  struct can_frame *rxdesc; /* Pointer to the receive frame descriptor. */
+
+  /* MSS CAN composite message objects */
+
+  pmpfs_can_msgobject_t tx_msg; /* Pointer to the transmit message object */
+  pmpfs_can_msgobject_t rx_msg; /* Pointer to the receive message object */
+
+  /* Work queue entries */
+
+  struct work_s rxwork;   /* for deferring rx interrupt work to the wq */
+  struct work_s txdwork;  /* For deferring tx done interrupt work to the wq */
+  struct work_s pollwork; /* For deferring poll work to the wq */
+
+  mpfs_can_filterobject_t filter;  /* hardware and software filters */
+
+  struct net_driver_s dev;  /* Interface understood by the Nuttx network */
+} mpfs_can_instance_t;
+
+/* Driver memory pool */
+
+#ifdef CONFIG_MPFS_MSS_CAN0
+static mpfs_can_instance_t g_can0;
+
+static uint8_t g_tx_pool0[(sizeof(struct can_frame) + TIMESTAMP_SIZE) *
```

Comment:
It would be smart to allocate these as uint32_t buffers. Then you'll get can_frames aligned to 32-bit boundaries and can greatly simplify the copying of the frame data and other fields, as you know that they are all 32-bit aligned

Reply:
hmm, since i think the CAN payload len `dev->d_len` is in bytes so it's just more natural to use uint8_t buffer. I took the reference from other CAN drivers also.

Reply:
for example , https://github.com/tiiuae/nuttx/blob/67b95d0142f9d2b721db1ae24fb235b0130df0db/arch/arm/src/imxrt/imxrt_flexcan.c#L314

Reply:
Sure, it was just a suggestion! You should anyhow align to 32-bits, in case you want to optimsize accessing of data. You can also use e.g. "aligned_data(sizeof(uint32))"